x-pack/platform/plugins/shared/security/server/saved_objects/saved_objects_security_extension.ts

```diff
@@ -1021,6 +1101,114 @@ export class SavedObjectsSecurityExtension implements ISavedObjectsSecurityExten
+      {
+        namespace: params.namespace,
+        objects: params.objects,
+      },
+      action
+    );
+  }
+
+  async internalAuthorizeChangeAccessControl<A extends string>(
+    params: AuthorizeChangeAccessControlParams,
+    action: SecurityAction
+  ): Promise<CheckAuthorizationResult<A>> {
+    if (!this.typeRegistry) {
+      throw new Error('Type registry is not defined');
+    }
+
+    this.accessControlService.setUserForOperation(this.getCurrentUserFunc());
+    const namespaceString = SavedObjectsUtils.namespaceIdToString(params.namespace);
+    const { objects } = params;
+    const { auditAction, authzAction } = this.decodeSecurityAction(action);
+
+    this.assertObjectsArrayNotEmpty(objects, action);
+
+    const objectsNotSupportingAccessControl = objects.every(
+      ({ type }) => !this.typeRegistry?.supportsAccessControl(type)
+    );
+
+    if (objectsNotSupportingAccessControl) {
+      const errMessage = `Unable to ${authzAction} for types ${[
+        ...objects.map(({ type }) => type),
+      ].join(', ')}`;
+      throw SavedObjectsErrorHelpers.decorateBadRequestError(new Error(errMessage));
+    }
+
+    const spacesToAuthorize = new Set<string>([namespaceString]);
+
+    const { typesRequiringAccessControl } =
+      this.accessControlService.getTypesRequiringPrivilegeCheck({
+        objects,
+        actions: new Set([action]),
+      });
+
+    /**
+     * AccessControl operations do not fall under the regular authorization actions for
+     * Saved Objects, but still require authorization. Hence, we pass an empty actions list to the base
+     * authorization checks.
+     */
+
+    if (typesRequiringAccessControl.size > 0) {
+      const authorizationResult = await this.checkAuthorization({
+        types: new Set(typesRequiringAccessControl),
+        spaces: spacesToAuthorize,
+        actions: new Set<A>([]),
+        options: { allowGlobalResource: true, typesRequiringAccessControl },
+      });
+
+      /**
+       * enforceAccessControl acts only on the current space and is only either fully_authorized or unauthorized.
+       * Even though this is a bulk operation, we don't allow partial changes to access control, i.e the incoming
+       * list of SOs must all support access control.
+       */
+      this.accessControlService.enforceAccessControl({
+        typesRequiringAccessControl,
+        authorizationResult,
+        currentSpace: namespaceString,
+      });
+
+      if (authorizationResult.status !== 'fully_authorized') {
+        const errMessage = `Unable to ${authzAction} for types ${[
+          ...typesRequiringAccessControl,
+        ].join(', ')}`;
+        const err = new Error(errMessage);
+        this.addAuditEvent({
+          action: auditAction!,
+          error: err,
+          unauthorizedTypes: [...typesRequiringAccessControl],
+          unauthorizedSpaces: [...spacesToAuthorize],
+        });
+        throw SavedObjectsErrorHelpers.decorateForbiddenError(err);
+      }
```

Comment:
Didn't we determine that this was unreachable and the audit log should be moved into the enforce call? Are going to handle this in the audit logging follow-up task?